htlcswitch/quiescer.go

```diff
@@ -0,0 +1,237 @@
+//nolint:unused
+package htlcswitch
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/fn"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// ErrStfuAlreadySent indicates that this channel has already sent an
+	// Stfu message for this negotiation.
+	ErrStfuAlreadySent = fmt.Errorf("stfu already sent")
+
+	// ErrStfuAlreadyRcvd indicates that this channel has already received
+	// an Stfu message for this negotiation.
+	ErrStfuAlreadyRcvd = fmt.Errorf("stfu already received")
+
+	// ErrNoDownstreamLeader indicates that the caller has requested the
+	// quiescence leader for a channel that is not yet quiescent.
+	ErrNoDownstreamLeader = fmt.Errorf(
+		"indeterminate quiescence leader: channel is not quiescent",
+	)
+
+	// ErrPendingRemoteUpdates indicates that we have received an Stfu while
+	// the remote party has issued updates that are not yet bilaterally
+	// committed.
+	ErrPendingRemoteUpdates = fmt.Errorf(
+		"stfu received with pending remote updates",
+	)
+
+	// ErrPendingLocalUpdates indicates that we are attempting to send an
+	// Stfu while we have issued updates that are not yet bilaterally
+	// committed.
+	ErrPendingLocalUpdates = fmt.Errorf(
+		"stfu send attempted with pending local updates",
+	)
+)
+
+type quiescerCfg struct {
+	// chanID marks what channel we are managing the state machine for. This
+	// is important because the quiescer is responsible for constructing the
+	// messages we send out and the ChannelID is a key field in that
+	// message.
+	chanID lnwire.ChannelID
+
+	// channelInitiator indicates which ChannelParty originally opened the
+	// channel. This is used to break ties when both sides of the channel
+	// send Stfu claiming to be the initiator.
+	channelInitiator lntypes.ChannelParty
+
+	// numPendingUpdates is a function that returns the number of pending
+	// originated by the party in the first argument that have yet to be
+	// committed to the commitment transaction held by the party in the
+	// second argument.
+	numPendingUpdates func(lntypes.ChannelParty,
```

Comment:
lol I highly highly disagree coming from an FP background. In this case we need *some* means of coordinating when the valid times to send an stfu message. Either the link needs to pass context to the quiescer, or the link needs to intermediate the process. The bottom line is that the quiescer and the link need to exchange info to work properly. This design is the evolution of @carlaKC and @Roasbeef 's recs.

Reply:
I think one reason to reach for slim interfaces at times over function closures is that interfaces play nicer with testing using the `mock` package (line up expectations, intercept behavior, etc).

Reply:
I would advocate for tying functions together via interface if they were functions that could properly relate to one another. In this case the functions in question are a function that can query the number of pending updates on the channel and another one that can send a message. There isn't really a way to tie these two together with any sort of invariant. So at minimum this would need to be two separate interfaces with a single method each, and I don't see that as an improvement over the functions themselves.

Reply:
> Too bad go is not haskell🙃

It's not so much a Go vs. Haskell thing. The problem that needs to be resolved is that the quiescer needs to be able to ask the question "can i send stfu yet?" in the current design because we made the quiescer responsible for actually sending stfu.

Given that, what language construct do we want to use to have it ask that question? Functions are the most natural way to do this in my experience. When we have collections of functions that all relate to each other in a well-defined and predictable way, we may bundle them into interfaces. Both Go and Haskell support both constructs: Go allows for function passing directly (as exemplified), and Haskell has its own version of interfaces (typeclasses).

To make the statement that "passing functions implies bad design" is what I'm challenging. The question I would ask at that point is... why is it bad design? Is it simply that we're using functions instead of interfaces? Is it that the quiescer shouldn't be responsible for sending the stfu message directly?

> Thus I don't see the benefit of creating aliases here, as I always prefer being explicit.

I'm not sure I completely follow what this is saying. I don't think there's any aliasing going on. We're just supplying the quiescer with a means of answering the question it needs to ask. What does explicit mean here?

> Think this is pretty minor so non-blocking.

Non-blocking or not I'd like to understand what the "correct" or at least "more correct" approach is in your mind. What's the concrete ask here? To me functions are the only obvious choice, but I could be missing some alternative way of thinking about the problem.

Reply:
To me passing methods usually ends up like [this](https://github.com/lightningnetwork/lnd/blob/f34bf51d164fe0472f1c0e354ae38f20ba9fd490/server.go#L688-L700) since there's no consensus about what rules to follow here (nor should we try to come up with one) - can we mutate state there? can we query states from multiple state machines? can we add locks inside it? I mean, it's a function so we are free to do anything here.

Now, what if we avoid passing methods - then the only option is to use interfaces or concrete structs. In this case, we'd have a `peer lnpeer.Peer` and a `channel *lnwallet.LightningChannel` placed in `quiscerCfg`, which is more explicit if we compare `q.cfg.sendMsg` vs `q.cfg.peer.SendMsg(false, stfu)` - I can easily tell it's dealing with a `Peer`, and sending the message is non-blocking due to `false`. But then we don't like the large interface or fat struct being attached to the cfg, and we don't need to stop here - notice the big concrete `*lnwallet.LightningChannel` struct, it should be an interface, whose responsibility is to answer questions about the channel state, so maybe there should be a small interface dedicated to that. Also the `lnpeer.Peer` interface is kinda large, maybe there should be an interface that only deals with `SendMsg`. And to me, all these possible improvements won't be attempted if we use methods as config values - it's already small so we could care less, while we are giving up imo a better design.


> Is it that the quiescer shouldn't be responsible for sending the stfu message directly?

Good q, which got me to think about the message streams used in the link - so we have `mailbox`, `upstream` and `downstream`, I thought we'd put msgs in the `mailbox` and let the switch dispath them, but apparently we are sending msgs directly in the link already, maybe you have an answer?

Reply:
> then the only option is to use interfaces or concrete structs

At the end of the day both of these are just collections of functions anyway. A struct completely fixes the choice of methods and an interface merely binds the choice of functions together.

> In this case, we'd have a peer lnpeer.Peer and a channel *lnwallet.LightningChannel placed in quiscerCfg

Doing this starts to create circular references in the object graph which I think is quite bad for being able to reason about the system. If there are no circular references then call graphs are guaranteed to bottom out at some point. In the presence of circular refs, you lose this guarantee. Secondly, if you go this route, you are giving access to *all* of that interfaces functionality which can cause further entanglement over time. Someone will inevitably use another method on the interface somewhere down the line simply because it is there without giving thought to whether or not it should be made available.

> Also the lnpeer.Peer interface is kinda large, maybe there should be an interface that only deals with SendMsg

The base case of this approach is that you end up with functions again. We can see in this case that I am using a single method on each of these interfaces/structures. Putting that method on an interface affords no benefit over just supplying the function. 

> And to me, all these possible improvements won't be attempted if we use methods as config values

I think this entire discussion hinges on this claim, which I find difficult to understand. Why is it that you think that choosing the function route will prevent improvements from being attempted? To me, it seems that it would be more likely since the lack of bundling of the functions permits a more flexible and incremental improvement strategy than an approach that involves a coincidental bucket of functions that don't have any relation to one another.

> maybe you have an answer?

Previous versions of this PR made it such that the burden of coordination happened inside the link which @carlaKC pushed against. In that design both sends and channel state queries were performed in the link and the quiescer's *only* job was to provide a queryable automata for the stfu process. Her chief complaint about this design is that it burdens the link with quiescence details, and that claim is correct. Given that there are coordination requirements that span both the channel state machine and the quiescence state machine, we only have two choices. Either we make the coordination happen in the link, or we make it happen in the quiescer. Whoever is responsible for doing that coordination needs to query the other machine to make sure that coordination is happening properly.

Either the link queries the quiescer for its state and then makes decisions and performs actions based off of that state, or the quiescer queries the link for its state and then makes decisions and performs actions based off of that. To me, the more logically sound and principled design is the original (the link queries the quiescer), however, the current design is cleaner to read in many cases (by hiding details, which also makes it tougher as you point out to figure out how this part of the process works).

Reply:
> A struct completely fixes the choice of methods and an interface merely binds the choice of functions together.

I prolly gonna scream on this lol. Nope, a struct is the implementation and an interface is an API design - we haven't spent enough time on the design part, but once we have small and deep interfaces, things would be way more easier.

> Doing this starts to create circular references...

Exactly! And by passing methods (not functions) we just hide these issues and possibly never have this kinda discussion.

> The base case of this approach is that you end up with functions again. We can see in this case that I am using a single method on each of these interfaces/structures. Putting that method on an interface affords no benefit over just supplying the function.

The benefit is we don't pretend we've solved the architecture issues by passing a wrapper around a method and think the work is done. Gonna pause the discussion here and maybe we could pick it up during calls, but my points are,

1) more explicit on the callsite
2) fits better in unit tests by using `mock`
3) drives design-wise improvements

Guess I wanna put interfaces everywhere.

> Previous versions of this PR made it such that the burden of coordination happened inside the link

More in the comments, but I think it's up to the `link` to decide whether we wanna turn over to the `quiescer` state machine.

Reply:
Resolved out of band. For now we will leave this design as is and work towards redesigning the link in the future. @yyforyongyu please thumbs up this comment and resolve to confirm.